src/librustc_trans/trans/_match.rs

```diff
@@ -1677,8 +1696,22 @@ fn bind_irrefutable_pat<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
 
                             ast::BindByRef(_) => {
                                 // By ref binding: the value of the variable
-                                // is the pointer `val` itself.
-                                Store(bcx, val, llval);
+                                // is the pointer `val` itself or fat pointer referenced by `val`
+                                if let ty::ty_rptr(_, ty::mt { ty: inner_ty, .. }) = ty.sty {
```

Comment:
`store_ty`? It looks like it is not supposed to be used with fat pointers (or non-immediate types in general).